<div class="section" lang="en" xml:lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="ch12lvl1sec76"></a>How does testing work?</h2></div></div><hr /></div><p>Let's start with a very simple Python function for us to test:</p><pre class="programlisting">def square(x): 
    return x * x </pre><p>In order to verify the correctness of this code, we pass a value, and will test if the result of the function is what we expect. For example, we could give it an input of 5, and would <span>expect</span><a id="id324974909" class="indexterm"></a> the result to be 25.</p><p>To illustrate the concept, we can manually test this function in the command line using the <code class="literal">assert</code> statement. The <code class="literal">assert</code> statement in Python simply says that if the conditional statement after the <code class="literal">assert</code> keyword returns <code class="literal">False</code>, then it will throw an exception as follows:</p><pre class="programlisting"><span class="strong"><strong>$ python</strong></span><span class="strong"><strong>&gt;&gt;&gt; def square(x): </strong></span><span class="strong"><strong>...     return x * x</strong></span><span class="strong"><strong>&gt;&gt;&gt; assert square(5) == 25</strong></span><span class="strong"><strong>&gt;&gt;&gt; assert square(7) == 49</strong></span><span class="strong"><strong>&gt;&gt;&gt; assert square(10) == 100</strong></span><span class="strong"><strong>&gt;&gt;&gt; assert square(10) == 0</strong></span>
    Traceback (most recent call last):
      File "&lt;stdin&gt;", line 1, in &lt;module&gt;
    AssertionError
  </pre><p>Using these <code class="literal">assert</code> statements, we verified that the square function was working as intended.</p><p>Â </p></div>